crates/red_knot_python_semantic/src/types/type_ordering.rs

```diff
@@ -264,11 +269,44 @@ pub(super) fn union_elements_ordering<'db>(left: &Type<'db>, right: &Type<'db>)
         (Type::Dynamic(_), _) => Ordering::Less,
         (_, Type::Dynamic(_)) => Ordering::Greater,
 
-        (Type::Union(left), Type::Union(right)) => left.cmp(right),
+        (Type::Union(left), Type::Union(right)) => {
+            if left.elements(db).len() != right.elements(db).len() {
+                return left.elements(db).len().cmp(&right.elements(db).len());
+            }
+            // Lexicographically compare the elements of the two unions.
+            for (left, right) in left.elements(db).iter().zip(right.elements(db)) {
+                let ordering = union_elements_ordering(left, right, db);
+                if ordering != Ordering::Equal {
+                    return ordering;
+                }
+            }
+            Ordering::Equal
+        }
         (Type::Union(_), _) => Ordering::Less,
         (_, Type::Union(_)) => Ordering::Greater,
 
-        (Type::Intersection(left), Type::Intersection(right)) => left.cmp(right),
+        (Type::Intersection(left), Type::Intersection(right)) => {
+            // Lexicographically compare the elements of the two intersections.
+            if left.positive(db).len() != right.positive(db).len() {
+                return left.positive(db).len().cmp(&right.positive(db).len());
+            }
+            if left.negative(db).len() != right.negative(db).len() {
+                return left.negative(db).len().cmp(&right.negative(db).len());
+            }
+            for (left, right) in left.positive(db).iter().zip(right.positive(db)) {
+                let ordering = union_elements_ordering(left, right, db);
+                if ordering != Ordering::Equal {
+                    return ordering;
+                }
+            }
+            for (left, right) in left.negative(db).iter().zip(right.negative(db)) {
+                let ordering = union_elements_ordering(left, right, db);
+                if ordering != Ordering::Equal {
+                    return ordering;
+                }
+            }
```

Comment:
I'm not convinced this is the source of the incremental-check regression (since I don't remember seeing the regression on earlier versions of the diff that included this code), but I do think it's probably worth extracting `{left,right}.negative(db)` and `{left,right}.positive(db)` into local variables so we call those queries just once each.

Reply:
> I'm not convinced this is the source of the incremental-check regression

FWIW, while there's still a regression on codspeed, it does look like https://github.com/astral-sh/ruff/pull/16641/commits/9718536fb866598b95b868c69db4265149770888 _significantly_ improved performance: https://codspeed.io/astral-sh/ruff/branches/mtshiba%3A14899-gradual-type-assignability